WKSnapshotConfiguration should have an SPI option to exclude selection highlighting

jeffmapple opened the pull request:
#### de25a61cb481e50d2c33495ed974079dd7462962
<pre>
WKSnapshotConfiguration should have an SPI option to exclude selection highlighting
<a href="https://bugs.webkit.org/show_bug.cgi?id=247634">https://bugs.webkit.org/show_bug.cgi?id=247634</a>
&lt;rdar://99892167&gt;

Reviewed by Tim Horton.

Add a _includesSelectionHighlighting property to WKSnapshotConfiguration on macOS. This controls
whether -[WKWebView takeSnapshotWithConfiguration:completionHandler:] captures the snapshot using
SnapshotOptionsExcludeSelectionHighlighting. By default, this property is YES to maintain
compatibility with existing clients.

* Source/WebKit/UIProcess/API/Cocoa/WKSnapshotConfiguration.mm:
(-[WKSnapshotConfiguration init]):
Initialize _includesSelectionHighlighting.

(-[WKSnapshotConfiguration copyWithZone:]):
Copy _includesSelectionHighlighting.

(-[WKSnapshotConfiguration _includesSelectionHighlighting]):
Added. This property can&apos;t be synthesized because it&apos;s defined in a category.

(-[WKSnapshotConfiguration _setIncludesSelectionHighlighting:]):
Ditto.

* Source/WebKit/UIProcess/API/Cocoa/WKSnapshotConfigurationPrivate.h:
Added.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView takeSnapshotWithConfiguration:completionHandler:]):
Optionally pass SnapshotOptionsExcludeSelectionHighlighting to takeSnapshot().

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Added WKSnapshotConfigurationPrivate.h.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSnapshot.mm:
(TEST):
Added a test that displays selected text and verifies that a snapshot can be captured without
selection highlighting.

Canonical link: <a href="https://commits.webkit.org/256592@main">https://commits.webkit.org/256592@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc144db726678e3b678a69b61586cd018447baaa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96255 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5510 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29305 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105801 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5631 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/34263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88632 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/102539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101926 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82851 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31195 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39987 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37664 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/20805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4571 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/83 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/91 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/40072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/UIProcess/API/Cocoa/WKSnapshotConfiguration.mm:
```diff
@@ -24,9 +24,13 @@
 
-@implementation WKSnapshotConfiguration
+@implementation WKSnapshotConfiguration {
+#if !TARGET_OS_IPHONE
```
This is an implementation file, so !PLATFORM(IOS_FAMILY) probably?

hortont424 replied:
(And below)

jeffmapple replied:
I'm going to go with `PLATFORM(MAC)` instead.

jeffmapple committed:
WKSnapshotConfiguration should have an SPI option to exclude selection highlighting
https://bugs.webkit.org/show_bug.cgi?id=247634
<rdar://99892167>

Reviewed by Tim Horton.

Add a _includesSelectionHighlighting property to WKSnapshotConfiguration on macOS. This controls
whether -[WKWebView takeSnapshotWithConfiguration:completionHandler:] captures the snapshot using
SnapshotOptionsExcludeSelectionHighlighting. By default, this property is YES to maintain
compatibility with existing clients.

* Source/WebKit/UIProcess/API/Cocoa/WKSnapshotConfiguration.mm:
(-[WKSnapshotConfiguration init]):
Initialize _includesSelectionHighlighting.

(-[WKSnapshotConfiguration copyWithZone:]):
Copy _includesSelectionHighlighting.

(-[WKSnapshotConfiguration _includesSelectionHighlighting]):
Added. This property can't be synthesized because it's defined in a category.

(-[WKSnapshotConfiguration _setIncludesSelectionHighlighting:]):
Ditto.

* Source/WebKit/UIProcess/API/Cocoa/WKSnapshotConfigurationPrivate.h:
Added.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView takeSnapshotWithConfiguration:completionHandler:]):
Optionally pass SnapshotOptionsExcludeSelectionHighlighting to takeSnapshot().

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Added WKSnapshotConfigurationPrivate.h.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSnapshot.mm:
(TEST):
Added a test that displays selected text and verifies that a snapshot can be captured without
selection highlighting.

Canonical link: https://commits.webkit.org/256592@main
